File: src/ruby/tormenta20/models/classe.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Classe (Character Class) in Tormenta20.
    #
    # Classes define the character's abilities, proficiencies, and progression.
    #
    # @example Find all spellcasting classes
    #   Classe.conjuradores
    #
    # @example Get class HP information
    #   guerreiro = Classe.find("guerreiro")
    #   guerreiro.initial_hp    # => 20
    #   guerreiro.hp_per_level  # => 5
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "guerreiro")
    # @!attribute [rw] name
    #   @return [String] Class name (e.g., "Guerreiro")
    # @!attribute [rw] hit_points
    #   @return [Hash] Hit points data with "initial" and "per_level" keys
    # @!attribute [rw] mana_points
    #   @return [Hash] Mana points data with "per_level" key
    # @!attribute [rw] skills
    #   @return [Hash] Skills data with "mandatory", "choose_amount", "choose_from" keys
    # @!attribute [rw] proficiencies
    #   @return [Hash] Proficiencies with "weapons", "armors", "shields" keys
    # @!attribute [rw] spellcasting
    #   @return [Hash, nil] Spellcasting info or nil if class cannot cast spells
    class Classe < Base
      self.table_name = "classes"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true

      # @!method conjuradores
      #   Filter classes that can cast spells.
      #   @return [ActiveRecord::Relation<Classe>]
      scope :conjuradores, -> { where.not(spellcasting: nil) }

      # Get initial hit points for this class.
      #
      # @return [Integer] Initial HP at level 1
      def initial_hp
        hit_points&.dig("initial") || 0
      end

      # Get hit points gained per level.
      #
      # @return [Integer] HP gained each level
      def hp_per_level
        hit_points&.dig("per_level") || 0
      end

      # Get mana points gained per level.
      #
      # @return [Integer] MP gained each level
      def mp_per_level
        mana_points&.dig("per_level") || 0
      end

      # Get mandatory skills for this class.
      #
      # @return [Array<String>] List of mandatory skill names
      def mandatory_skills
        skills&.dig("mandatory") || []
      end

      # Get number of skills to choose from available list.
      #
      # @return [Integer] Number of skills to choose
      def choose_skills_amount
        skills&.dig("choose_amount") || 0
      end

      # Get skills available to choose from.
      #
      # @return [Array<String>] List of available skill names
      def available_skills
        skills&.dig("choose_from") || []
      end

      # Get weapon proficiencies for this class.
      #
      # @return [Array<String>] List of weapon categories (e.g., ["simples", "marciais"])
      def weapon_proficiencies
        proficiencies&.dig("weapons") || []
      end

      # Get armor proficiencies for this class.
      #
      # @return [Array<String>] List of armor categories (e.g., ["leves", "pesadas"])
      def armor_proficiencies
        proficiencies&.dig("armors") || []
      end

      # Check if class has shield proficiency.
      #
      # @return [Boolean] true if proficient with shields
      def shield_proficiency?
        proficiencies&.dig("shields") || false
      end

      # Check if this class can cast spells.
      #
      # @return [Boolean] true if class has spellcasting ability
      def conjurador?
        !spellcasting.nil?
      end

      # Convert class to a hash representation.
      #
      # @return [Hash] Hash with all class attributes
      def to_h
        {
          id: id,
          name: name,
          hit_points: hit_points,
          mana_points: mana_points,
          skills: skills,
          proficiencies: proficiencies,
          abilities: abilities,
          powers: powers,
          progression: progression,
          spellcasting: spellcasting
        }.compact
      end
    end
  end
end

File: src/ts/models/classe.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "classes"

/**
 * A Classe (character class) in Tormenta20.
 *
 * @example
 * ```ts
 * Classe.all()                    // all classes
 * Classe.conjuradores().all()     // spellcasting classes only
 * Classe.find("guerreiro")
 *
 * const guerreiro = Classe.find("guerreiro")!
 * guerreiro.initialHp             // 20
 * guerreiro.hpPerLevel            // 5
 * guerreiro.isConjurador          // false
 * ```
 */
export class Classe extends BaseModel {
  /** Raw hit_points JSON object with `initial` and `per_level` keys. */
  get hitPoints(): Record<string, number> { return BaseModel.json(this._row.hit_points, { initial: 0, per_level: 0 }) }
  /** Raw mana_points JSON object with `per_level` key. */
  get manaPoints(): Record<string, number> { return BaseModel.json(this._row.mana_points, { per_level: 0 }) }
  /** Raw skills JSON with `mandatory`, `choose_amount`, and `choose_from` keys. */
  get skills(): Record<string, unknown> { return BaseModel.json(this._row.skills, {}) }
  /** Raw proficiencies JSON with `weapons`, `armors`, and `shields` keys. */
  get proficiencies(): Record<string, unknown> { return BaseModel.json(this._row.proficiencies, {}) }
  /** Class ability objects (level-up features). */
  get abilities(): unknown[] { return BaseModel.json(this._row.abilities, []) }
  /** Class power unlocks by level. */
  get powers(): unknown[] { return BaseModel.json(this._row.powers, []) }
  /** Level-by-level progression table rows. */
  get progression(): unknown[] { return BaseModel.json(this._row.progression, []) }
  /**
   * Spellcasting data (attribute, circle progression, etc.), or `null` for non-casters.
   *
   * Check {@link isConjurador} for a boolean shortcut.
   */
  get spellcasting(): Record<string, unknown> | null {
    const v = BaseModel.json<Record<string, unknown> | null>(this._row.spellcasting, null)
    return v && Object.keys(v).length > 0 ? v : null
  }

  /** HP at level 1. */
  get initialHp(): number { return this.hitPoints.initial ?? 0 }
  /** HP gained per level. */
  get hpPerLevel(): number { return this.hitPoints.per_level ?? 0 }
  /** MP gained per level. */
  get mpPerLevel(): number { return this.manaPoints.per_level ?? 0 }
  /** Skills all members of this class are trained in. */
  get mandatorySkills(): unknown[] { return (this.skills.mandatory as unknown[]) ?? [] }
  /** Number of skills the player may choose at character creation. */
  get chooseSkillsAmount(): number { return (this.skills.choose_amount as number) ?? 0 }
  /** Pool of skills the player chooses from. */
  get availableSkills(): unknown[] { return (this.skills.choose_from as unknown[]) ?? [] }
  /** Weapon category proficiencies (e.g. `["simples", "marciais"]`). */
  get weaponProficiencies(): string[] { return (this.proficiencies.weapons as string[]) ?? [] }
  /** Armor category proficiencies (e.g. `["leves"]`). */
  get armorProficiencies(): string[] { return (this.proficiencies.armors as string[]) ?? [] }
  /** Whether the class is proficient with shields. */
  get shieldProficiency(): boolean { return !!(this.proficiencies.shields) }
  /** `true` if the class can cast spells (has non-empty spellcasting data). */
  get isConjurador(): boolean { return this.spellcasting !== null }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, hit_points: this.hitPoints, mana_points: this.manaPoints,
      skills: this.skills, proficiencies: this.proficiencies, abilities: this.abilities,
      powers: this.powers, progression: this.progression, spellcasting: this.spellcasting,
    }
  }

  static query(): Query<Classe> { return BaseModel.makeQuery(TABLE, Classe) }
  static all(): Classe[] { return Classe.query().all() }
  static find(id: string): Classe | null { return Classe.query().find(id) }
  static first(): Classe | null { return Classe.query().first() }
  static last(): Classe | null { return Classe.query().last() }
  static count(): number { return Classe.query().count() }
  /** Returns only classes that have spellcasting capability. */
  static conjuradores(): Query<Classe> { return Classe.query().where("spellcasting IS NOT NULL AND spellcasting != '{}'") }
}
